### PR TITLE
Wasm compatible

### PR DESCRIPTION
**Pull request by gustav0frings**
Current implementation cannot be compiled to wasm because it uses std features and also openssl.

This new implemention has dependencies that don rely on std features. Usage of vectors, format, println and other macros has been removed as well.
I used #![no_std] attribute and I made sure that this crate compiles in wasm in its current version.

The code has also been refactored and made more simple. 
I created a full module and structured the crate to be used externally.

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -2,450 +2,471 @@
...
+        _ => panic!("Failed to decode CBOR payload:{:?}", payload),
+    };
+
+    let certificate = payload
+        .get(&serde_cbor::Value::Text("certificate".try_into().unwrap()))
+        .expect("certificate not found");
+
+    //println!("certificate: {:?}", certificate);
+
+    let certifcate_bytes: [u8; 643] = serde_cbor::to_vec(&certificate)
```
Comment (PratyushRT) [inline src/lib.rs]: Lets fix this typo in naming, variable name is consistent but just for quality

**Review by PratyushRT — CHANGES_REQUESTED** [review]
Overall lots of solid work in here @asa93 the COSE parsing, X.509 extraction, signature verification all looks good. 

Lets address the following aspects, we can also discuss any that might not be clear:

1. The code doesn't verify the certificate chain. It should validate the entire chain up to a trusted root certificate.
2. There's no check for the certificate's validity period: maybe this is something we do in a future version, or if easy lets add it. In the future we might even have to add certificate revocation checks (CRL or OCSP).
3. I think we should validate the extracted PCR against expected values.
4. The `expect()` in many places will panic on errors so we might want to handle it more gracefully.
5. Missing Checks: No verification of the module_id, no nonce checking 

A side note: looks like we are manually extracting public key from the certificate. Using a standard library function would be more robust. Was there nothing suitable in existing libraries?

**Review by saberistic — APPROVED** [review]
- [x] Parsing and Verifying of Attestation doc generated by Nitriding
- [x] Parsing all certificate related materials from the blob
- [x] Verifying x509 certificates with AWS root
- [x] Verifying attestation document signature
- [x] checking nonce
- [x] checking PCRs
- [x] Github actions with build/test and publish upon merging into main

**Review by PratyushRT — COMMENTED** [review]
Overall looks great, one small change that we can quickly discuss is considering using a custom error type instead of String for error handling. This would provide more structured and type-safe error handling. Something like the following:

```
use thiserror::Error;

#[derive(Error, Debug)]
pub enum VerificationError {
    #[error("Invalid nonce")]
    InvalidNonce,
    #[error("Invalid PCR at index {0}")]
    InvalidPCR(usize),
    #[error("X509 certificate verification failed: {0}")]
    X509CertVerificationFailed(String),
    #[error("Remote attestation signature verification failed: {0}")]
    SignatureVerificationFailed(String),
    // ... other error types ...
}

// Then update function signatures to use this error type
pub fn verify(...) -> Result<(), VerificationError>
```

**Comment by gustav0frings** [comment]
great job

**Comment by saberistic** [comment]
> great job

you did all the work, I was just the janitor

**Review by PratyushRT — APPROVED** [review]

**Pull request merged**